lib/twilio/audiohelper.ts

```diff
@@ -738,45 +751,51 @@ class AudioHelper extends EventEmitter {
-    }
+      const device: MediaDeviceInfo | undefined = this.availableInputDevices.get(deviceId);
+      if (!device) {
+        return Promise.reject(new InvalidArgumentError(`Device not found: ${deviceId}`));
+      }
 
-    this._log.info('Setting input device. ID: ' + deviceId);
+      this._log.info('Setting input device. ID: ' + deviceId);
 
-    if (this._inputDevice && this._inputDevice.deviceId === deviceId && this._selectedInputDeviceStream) {
-      if (!forceGetUserMedia) {
-        return Promise.resolve();
-      }
+      if (this._inputDevice && this._inputDevice.deviceId === deviceId && this._selectedInputDeviceStream) {
+        if (!forceGetUserMedia) {
+          return Promise.resolve();
+        }
 
-      // If the currently active track is still in readyState `live`, gUM may return the same track
-      // rather than returning a fresh track.
-      this._log.info('Same track detected on setInputDevice, stopping old tracks.');
-      this._stopSelectedInputDeviceStream();
-    }
+        // If the currently active track is still in readyState `live`, gUM may return the same track
+        // rather than returning a fresh track.
+        this._log.info('Same track detected on setInputDevice, stopping old tracks.');
+        this._stopSelectedInputDeviceStream();
+      }
 
-    // Release the default device in case it was created previously
-    this._stopDefaultInputDeviceStream();
+      // Release the default device in case it was created previously
+      this._stopDefaultInputDeviceStream();
 
-    const constraints = { audio: Object.assign({ deviceId: { exact: deviceId } }, this.audioConstraints) };
-    this._log.info('setInputDevice: getting new tracks.');
-    return this._getUserMedia(constraints).then((originalStream: MediaStream) => {
+      const constraints = { audio: Object.assign({ deviceId: { exact: deviceId } }, this.audioConstraints) };
+      this._log.info('setInputDevice: getting new tracks.');
+      return this._getUserMedia(constraints).then((originalStream: MediaStream) => {
 
-      this._destroyProcessedStream();
+        this._destroyProcessedStream();
 
-      return this._maybeCreateProcessedStream(originalStream).then((newStream) => {
-        this._log.info('setInputDevice: invoking _onActiveInputChanged.');
-        return this._onActiveInputChanged(newStream).then(() => {
-          this._replaceStream(originalStream);
-          this._inputDevice = device;
-          this._maybeStartPollingVolume();
+        return this._maybeCreateProcessedStream(originalStream).then((newStream) => {
+          this._log.info('setInputDevice: invoking _onActiveInputChanged.');
+          return this._onActiveInputChanged(newStream).then(() => {
+            this._replaceStream(originalStream);
+            this._inputDevice = device;
+            this._maybeStartPollingVolume();
+          });
         });
       });
+    };
+
+    return this._inputDevicePromise = setInputDevice().finally(() => {
```

Comment:
should we use
```
try {
} catch {
finally
```
for an error not caught in the `if` statements?

Reply:
The error is bubbled up all the way to the calling function and no need to catch here.